### PR TITLE
Move away from hand rolled migration to a migration tool

### PR DESCRIPTION
**Pull request by vagmi**
**Describe the technical changes contained in this PR**

We have been handrolling migrations and left it to the discretion of the developer to write a defensive migration. This will affect us in the long run. This PR implements the golang migrate tool.

https://github.com/golang-migrate/migrate

**Previous behaviour**
Migrations and structure changes were always rerun. Developers did not have a lot of control to migrate up or down.

**New behaviour**
With the migration tool a migration is run only once and can also be rolled back. We still use plain SQL for tracking migrations. Rather than timestamp based versions, we are using sequence based files just like the previous approaches.

**Have the following been addressed?**
- [x] Have test cases been created for all of the changes?
- [x] Do all of the test cases pass?
- [x] Has the testing been done using the default docker-compose environment?
- [x] Are documentation changes required?
- [x] Does this change break or alter existing behaviour?

### CODE REVIEW

File: backend/db_migrations/000001_common.up.sql [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    new.created_at := current_timestamp;
+    new.updated_at := new.created_at;
+    return new;
+  end;
+  $$ language plpgsql;
+
+  
+  -- nanoid imports from https://github.com/viascom/nanoid-postgresql
+  -- Apache 2.0 license ©️  Viascom
+  CREATE OR REPLACE FUNCTION nanoid(
```
Comment (shanthisa) [inline backend/db_migrations/000001_common.up.sql:69]: Write drop function in migration down

File: backend/db_migrations/000004_product.up.sql [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    product_keywords text[], 
+    tags text[],
+    created_at timestamp default now() not null,
+    updated_at timestamp default now() not null
+);
+
+create index if not exists 
+    idx_product_product_keywords on app.product using gin(product_keywords);
+
+comment on table app.product is 'Product details for off-the-shelf ingredients. In a Meal this combines with the measure to describe the ingredient and portion. The Product list from Meals in a Meal Plan will be input to a Shopping List.';
```
Comment (shanthisa) [inline backend/db_migrations/000004_product.up.sql]: We no longer have measure. We need to mention about match here

File: backend/db_migrations/000002_auth_system.up.sql [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        new.role = 'app_admin';
+    end if;
+    return new;
+end;
+$$ language plpgsql security definer;
+create trigger tg_person_make_admin 
+before insert on app.person 
+for each row execute procedure app.make_first_user_admin();
+
+create or replace function app.authorize_admin(p_id bigint) returns app.person as $$
```
Comment (shanthisa) [inline backend/db_migrations/000002_auth_system.up.sql:116]: Do we need drop for authorize functions as other functions are dropped in down?

**Review by shanthisa — CHANGES_REQUESTED** [review]
Some cosmetic changes required

File: backend/db_migrations/000004_product.up.sql (hunk 2 above)
Comment (vagmi) [reply backend/db_migrations/000004_product.up.sql]: Done.

File: backend/db_migrations/000002_auth_system.up.sql (hunk 3 above)
Comment (vagmi) [reply backend/db_migrations/000002_auth_system.up.sql:116]: Done.

**Review by shanthisa — APPROVED** [review]
LGTM 👍

**Pull request merged**